### PR TITLE
[SPARK-36172][SS] Document session window into Structured Streaming guide doc

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR documents a new feature "native support of session window" into Structured Streaming guide doc.

Screenshots are following:

스크린샷 2021-07-20 오후 5 04 20

스크린샷 2021-07-20 오후 3 34 38

### Why are the changes needed?

This change is needed to explain a new feature to the end users.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Documentation changes.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @viirya @xuanyuanking

**Review by viirya — COMMENTED** [review]
Isn't w1 ending at 12:14 overlapping (1min) with w2 starting at 12:15 and w2 ending at 12:20 overlapping (2mins) w3 starting at 12:22 in the gap duration (5mins)?

**Review by viirya — COMMENTED** [review]
The doc looks okay, just a question about the diagram.

**Comment by HeartSaVioR** [comment]
Not sure I understand. w1 ends at 12:14 (12:09 + 5 mins) hence doesn't overlap with w2 which starts from 12:15. Vice versa for w2 and w3.

**Comment by viirya** [comment]
Oh, I misunderstood the diagram.

**Comment by viirya** [comment]
Maybe we can add a sentence explaining it too?

**Comment by HeartSaVioR** [comment]
OK we can probably add explanation for session end, as it looks to be not intuitive enough. Let me work on it. Thanks!

**Comment by HeartSaVioR** [comment]
Just updated the figure and also updated the screenshot of the PR.

**Review by viirya — APPROVED** [review]

**Review by viirya — COMMENTED** [review]
What `structured-streaming.pptx ` is used for?

**Comment by HeartSaVioR** [comment]
Looks like Spark project is doing great to source control everything including figures. Figures in doc seem to be created from Powerpoint and each PPTX file contains the figures per corresponding doc page.

**Comment by HeartSaVioR** [comment]
I'll merge this late tomorrow if there's no further comment.

File: docs/structured-streaming-programming-guide.md [hunk 1]
```diff
@@ -1063,6 +1063,77 @@ then drops intermediate state of a window < watermark, and appends the final
...
+can be bound to the multiple windows.
+
+Tumbling and sliding window use `window` function, which has been described on above examples.
+
+Session windows have different characteristic compared to the previous two types. Session window has a dynamic size
+of the window length, depending on the inputs. A session window starts with an input, and expands itself
+if following input has been received within gap duration. A session window closes when there's no input
+received within gap duration after receiving the latest input.
+
+Session window uses `session_window` function. The usage of the function is similar to the `window` function.
```
Comment (xuanyuanking) [inline docs/structured-streaming-programming-guide.md:1086]: Do we need to mention that the `session_window` function should be a grouping key and should not be the only grouping key?

**Review by xuanyuanking — APPROVED** [review]
LGTM.
Just an open question left.

File: docs/structured-streaming-programming-guide.md (hunk 1 above)
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md:1086]: Actually there's no such restriction, like we have no such restriction for `window` function. You can use `session_window` anywhere we expect expression. We just don't calculate any windowing aggregation if the function is not used in groping key, same as we do for `window` function.

It should not be the only grouping key <= this restriction is only applied to the streaming query. In batch query we allow global window.

Comment (xuanyuanking) [reply docs/structured-streaming-programming-guide.md:1086]: Ah yea make sense, I only considered the SS scenario.

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing! Merging to master/3.2

**Comment by HeartSaVioR** [comment]
Thanks. I merged this into master/3.2.

**Pull request closed without merging**